source/draw/gpu/PyramidImage.ooc

```diff
@@ -0,0 +1,166 @@
+import gpu/GpuYuv420Semiplanar
+import gpu/GpuMonochrome
+import Image
+
+PyramidImage: class extends Image {
+	_fullRepresentation: RasterMonochrome
+	_validRaster := false
+	fullRepresentation: RasterMonochrome {
+		get {
+			if (!this _validRaster)
+				this _toRaster()
+			this _fullRepresentation
+		}
+	}
+	_fullRepresentationGpu: GpuMonochrome
+	fullRepresentationGpu ::= this _fullRepresentationGpu
+
+	_base: RasterMonochrome
+	base: RasterMonochrome {
+		get {
+			if (this _base == null) {
+				this _base = this _getRepresentationSlice(0, this scaledResolution)
+				this _base referenceCount increase()
+			}
+			this _base
+		}
+	}
+	_representation: RasterMonochrome
+	representation: RasterMonochrome {
+		get {
+			if (this _representation == null) {
+				this _representation = this _getRepresentationSlice(this _scaledResolution area, IntSize2D new(this _scaledResolution width, this _representationRows))
+				this _representation referenceCount increase()
+			}
+			this _representation
+		}
+	}
+
+	//TODO: Not pretty, but I couldn't get this to work with any array type
+	_level1: RasterMonochrome
+	_level2: RasterMonochrome
+	_level3: RasterMonochrome
+	_level4: RasterMonochrome
+	getLevel: func (level: Int) -> RasterMonochrome {
+		result: RasterMonochrome
+		size := this _scaledResolution
+		offset := 0
+		for (i in 0..level) {
+			offset += size area
+			size /= 2
+		}
+		match(level) {
+			case 0 =>
+			 result = this base
+			case 1 =>
+				if (this _level1 == null) {
+					this _level1 = this _getRepresentationSlice(offset, size)
+					this _level1 referenceCount increase()
+				}
+				result = this _level1
+			case 2 =>
+				if (this _level2 == null) {
+					this _level2 = this _getRepresentationSlice(offset, size)
+					this _level2 referenceCount increase()
+				}
+				result = this _level2
+			case 3 =>
+				if (this _level3 == null) {
+					this _level3 = this _getRepresentationSlice(offset, size)
+					this _level3 referenceCount increase()
+				}
+				result = this _level3
+			case 4 =>
+				if (this _level4 == null) {
+					this _level4 = this _getRepresentationSlice(offset, size)
+					this _level4 referenceCount increase()
+				}
+				result = this _level4
+		}
+		result
+	}
+
+	_pyramidLevels: Int
+	pyramidLevels ::= this _pyramidLevels
+	_scaledResolution: IntSize2D
+	scaledResolution ::= this _scaledResolution
+	_representationRows: Int
+	representationRows ::= this _representationRows
+
+	init: func ~fromRepresentation (representation: GpuMonochrome, =_pyramidLevels, =_scaledResolution, =_representationRows) {
+		super(this _scaledResolution)
+		this _validRaster = false
+		if (this _fullRepresentationGpu != null) this _fullRepresentationGpu recycle()
+		this _fullRepresentationGpu = representation
+	}
+	_getRepresentationSlice: func(offset: Int, size: IntSize2D) -> RasterMonochrome {
+		result := RasterMonochrome new(this fullRepresentation buffer slice(offset, size area), size)
+		result
```

Comment:
Just `RasterMonochrome new(this fullRepresentation buffer slice(offset, size area), size)` would do nicely, there's no reason to use a `result` variable for one-liners.